sympy/solvers/ode/riccati.py

```diff
@@ -0,0 +1,506 @@
+to ``presol``.
+
+Step 13 : For each solution in ``presol``, apply an inverse transformation
+and append them to ``sol``, so that the solutions of the original equation
+are found using the solutions of the equation in its normal form.
+"""
+
+from sympy.core import S
+from sympy.core.numbers import oo
+from sympy.core.relational import Eq
+from sympy.core.symbol import symbols, Dummy
+from sympy.functions import sqrt
+from sympy.functions.elementary.complexes import sign
+from sympy.polys.domains import ZZ
+from sympy.polys.polytools import Poly
+from sympy.polys.polyroots import roots
+from sympy.solvers.solveset import linsolve
+
+def riccati_normal(w, x, b1, b2):
+    # y(x) = -b2(x)*w(x) - b2'(x)/(2*b2(x)) - b1(x)/2
+    return -b2*w - b2.diff(x)/(2*b2) - b1/2
+
+
+def riccati_inverse_normal(y, x, b1, b2, bp=None):
+    if bp is None:
+        bp = -b2.diff(x)/(2*b2**2) - b1/(2*b2)
+    # w(x) = -y(x)/b2(x) - b2'(x)/(2*b2(x)^2) - b1(x)/(2*b2(x))
+    return -y/b2 + bp
+
+
+def linsolve_dict(eq, syms, var):
+    sol = linsolve(eq, syms, var)
+    if not sol:
+        return {}
+    return {k:v for k, v in zip(syms, list(sol)[0])}
+
+
+def find_poles(num, den, x):
+    # All roots of denominator are poles of a(x)
+    p = roots(den, x)
+    # Substitute 1/x for x and check if oo is a pole
+    num, den = inverse_transform_poly(num, den, x)
+
+    # Find the poles of a(1/x)
+    p_inv = roots(den, x)
+
+    # If 0 is a pole of a(1/x), oo is a pole of a(x)
+    if 0 in p_inv:
+        p.update({oo: p_inv[0]})
+    return p
+
+
+def val_at_inf(num, den, x):
+    # Valuation of a rational function at oo = deg(denom) - deg(numer)
+    return den.degree(x) - num.degree(x)
+
+
+def check_necessary_conds(val_inf, muls):
+    return (val_inf >= 2 or (val_inf <= 0 and val_inf%2 == 0)) and all([mul == 1 or (mul%2 == 0 and mul >= 2) for mul in muls])
+
+
+def inverse_transform_poly(num, den, x):
+    """
+    A function to make the substitution
+    x -> 1/x in Poly objects.
+    """
+    # Declare for reuse
+    one = Poly(1, x)
+    xpoly = Poly(x, x)
+
+    # Check if degree of numerator is same as denominator
+    pwr = val_at_inf(num, den, x)
+    if pwr >= 0:
+        # Denominator has greater degree. Substituting x with
+        # 1/x would make the extra power go to the numerator
+        if num.expr != 0:
+            num = num.transform(one, xpoly) * x**pwr
+            den = den.transform(one, xpoly)
+    else:
+        # Numerator has greater degree. Substituting x with
+        # 1/x would make the extra power go to the denominator
+        num = num.transform(one, xpoly)
+        den = den.transform(one, xpoly) * x**(-pwr)
+    return num.cancel(den, include=True)
+
+
+def limit_at_inf(num, den, x):
+    pwr = -val_at_inf(num, den, x)
+    if pwr > 0:
+        return oo*sign(num.LC()/den.LC())
+    elif pwr == 0:
+        return num.LC()/den.LC()
+    else:
+        return 0
+
+
+def construct_c(num, den, x, poles, muls):
```

Comment:
I think this should be changed to just be a function that makes `c` for a single pole. At least it should call a function that works for a single pole. Then we can have functions for each step like:
```python
def get_c(...):
    if ...:
       return get_c_case1(...)
    ...

def get_c_case1(...):
    ...

def get_c_case2(...):
    ...
```
Each of these functions can have a docstring that explains the relevant concepts from the algorithm and the thesis.

It would be good to construct explicit simple examples of each case by hand and carefully write manual tests that verify each step of the algorithm that the correct values are returned for `c` and that any statements from the thesis about the form of the solution are correct.